Times of The Day Binary Sensor

kstaniek opened the pull request:
## Description:
This code provides the new binary sensor platform checking if the current time is within configured time range. The platform allows adding multiple sensor working for different time ranges.

The time ranges can be defined as time format string or sun event (sunset or sunrise). Also for each boundary the offset can be added.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8172

## Example entry for `configuration.yaml` (if applicable):
```yaml
binary_sensor:
  - platform: tod
    sensors:
      early_morning:
        after: sunrise
        after_offset: '-02:00'
        before: '07:00'
      morning:
        after: sunrise
        before: '12:00'
      late_morning:
        after: '10:00'
        before: '12:00'
      early_afternoon:
        after: '12:00'
        before: '15:00'
      afternoon:
        after: '12:00'
        before: '18:00'
      late_afternoon:
        after: '16:00'
        before: '18:0'
      early_evening:
        after: '17:00'
        before: '19:00'
      evening:
        after: '17:00'
        before: '22:00'
      late_evening:
        after: '20:00'
        before: '22:00'
      middle_of_the_night:
        after: '23:00'
        before: sunrise
        before_offset: '-02:00'
      night:
        after: '22:00'
        before: sunrise
      day:
        after: sunrise
        before: sunset
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

kstaniek committed:
First commit

kstaniek committed:
Times of the Day binary sensor added

dgomes reviewed (commented):
Just a very quick review.

I'm wondering what you couldn't do with sun.sun that you required this binary_sensor ?

kstaniek commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,223 @@
+            ENTITY_ID_FORMAT, f'tod_{sensor_name}', hass=hass)
+        self._calculate_initial_boudary_time()
+        _LOGGER.error(self._time_after.isoformat())
+        _LOGGER.error(self._time_before.isoformat())
```
Yes, that is a last time leftover. Removing.

kstaniek commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,223 @@
+        return False
+
+    async_add_entities(sensors)
+    return True
```
Removed.

kstaniek commented:
@dgomes 
> I'm wondering what you couldn't do with sun.sun that you required this binary_sensor ?

All cases you can implement with sun and binary_sensor.template however:
1. The template becomes more complex if you want to have a sun event with combination of offsets and take care of over midnight periods, i.e. from sunset to sunrise next day.
2. Sun component gives you only next event not today's.
3. Template sensor based on sun.sun may have unknown state during hass restart when sun.sun is not available yet. This is enjoying when you restart hass during the time period and your automation are triggered accidentally.

This sensor proposal aligns with one of the important @balloob 's goal, which is making hass simple to normal users.

I have also an idea to add another sensor.tod with values calculated based of different non-linear functions. I.e. norma distribution with float values from 0 to 1. This feature may be extremely
helpful in ML where you need time quantification and classificators.

Please let me know if this convinces you and I will make a fixed pull request.
Thanks,
Klaus

kstaniek committed:
Python 3.5 fixes and logging removed

OttoWinter commented:
For config schema: https://github.com/home-assistant/architecture/issues/12#issuecomment-366548001

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+        vol.Lower, vol.Any(cv.sun_event))),
+    vol.Optional(CONF_AFTER_OFFSET, default=timedelta(0)): cv.time_period,
+    vol.Required(CONF_BEFORE): vol.Any(cv.time, vol.All(
+        vol.Lower, vol.Any(cv.sun_event))),
```
Why this `vol.Any` here?

kstaniek replied:
No reason. Previously had a list there but realised that cv.sunevent is pre_defined. Remove.

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+        self._next_sunsetting = None
+
+        self.entity_id = async_generate_entity_id(
+            ENTITY_ID_FORMAT, 'tod_{}'.format(sensor_name), hass=hass)
```
Let's use the built-in entity id generator. Prefixing the entity is by `tod_` is unexpected.

kstaniek replied:
You mean just pass the sensor_name to generate_entity_id without prepending with tod_. Makes sense to avoid confusion as the name is provided in config.

OttoWinter replied:
Well now that you're using the default entity ID format, there's no need for adding this explicitly. Home Assistant will automatically generate the entity ID for you.

Please remove `async_generate_entity_id` here. It will automatically be set.

kstaniek replied:
OK, Understand removed async_generate_entity_id and replaced with: 
`self.entity_id = ENTITY_ID_FORMAT.format(name)`

OttoWinter replied:
No, I mean removing this entire line. HA will automatically set the entity ID, no need to assign it yourself

kstaniek replied:
Why? I have the sensor ids in the config. I don't want entity_id to be set from friendly name? Is that wrong?

```
binary_sensor:
  - platform: tod
    sensors:
      early_morning:  # I want this to be entity Id 
        after: sunrise
        after_offset: '-02:00'
        before: '07:00'
        friendly_name: "Wszesny poranek"
      morning:
        after: sunrise
        before: '12:00'
```

OttoWinter replied:
Yes it is correct for template binary sensors. However, we decided some time ago to move to a new configuration format for new platforms like this. Please see the other comments I've left here, for example this one: https://github.com/home-assistant/home-assistant/pull/20068#issuecomment-453919773

kstaniek replied:
Is that the correct comment it?

kstaniek replied:
@OttoWinter Can you have a look, please:

```
1)

binary_sensor:
  - platform: tod
    sensors:
      early_morning:
        after: sunrise
        after_offset: '-02:00'
        before: '07:00'
        friendly_name: "Wszesny poranek"   # used local language on purpose
      morning:
        after: sunrise
        before: '12:00'
2)

binary_sensor:
  - platform: tod
    tods:
      - name: Wczesny poranek  # the entity_id will be automatically generated based on this
        after: sunrise
        after_offset: '-02:00'
        before: '07:00'
      - name: Morning
        after: sunrise
        before: '12:00'
```

So is my understanding correct you would recommend option 2 for that sensor?
IMHO option 1 is much easier to understand when you read it. It clearly indicates that there are multiple sensors created with this platforms. Also If English is not your native language and you still want to keep your entity_id in English, for sake of consistency with the rest of you config, and friendly_name in local language, option 2 seems be better than generating the entity id based on the name which may be in other language.
I don't have enough experience yet, and maybe I can't see issues that are obvious to you. 
Would appreciate your comment and explanation. I'm not tight to any option yet, and willing to get into final consensus here and eventually close this PR.

kstaniek replied:
@OttoWinter Hi Otto, could you please take a look at my comments?
There could be also a third option:

```
3) 
binary_sensor:
  - platform: tod
     name: Wczesny poranek
     after: sunrise
     after_offset: '-02:00'
     before: '07:00'
  
  - platform: tod 
     name: Morning
     after: sunrise
     before: '12:00'
```

I'm a bit confused and It's not clear to me what option is preferred.

kstaniek replied:
@OttoWinter Any reply on this pls?

OttoWinter replied:
@kstaniek Yes, option 3. The template sensor way of doing it all in `sensors:` was always very confusing to users (from community forum issue count)

kstaniek replied:
Updated the config and tests accordingly. Option 3 has been implemented for config.

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+    return event in (SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET)
+
+
+class TodSensor(Entity):
```
This should inherit from `BinarySensorDevice`

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+        if self.after < self.before:
+            if self.after <= self.current_datetime < self.before:
+                return STATE_ON
+        return STATE_OFF
```
Don't override state, override is_on

kstaniek replied:
ACK

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+                 before, before_offset):
+        """Init the ToD Sensor..."""
+        self._name = sensor_name
+        self._state = None
```
This is never used

kstaniek replied:
Yes, true. Remove.

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+        return
+
+    async_add_entities(sensors)
+    return
```
Unnecessary return

kstaniek replied:
OK, but explicit is better than implicit ;-)

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+
+    if not sensors:
+        _LOGGER.error("No sensors added")
+        return
```
This should be handled by the validation step.

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+                self._time_before += timedelta(days=1)
+
+        self._time_after += self._after_offset
+        self._time_before += self._before_offset
```
I think there could be issues with daylight savings time here. For example if you add 1 day to March 30th in CET at 2:30, the result will be a time that does not exist! So, depending on how `async_track_point_in_time` works, this will never fire.

kstaniek replied:
That's something that need to solve and add the test case.

kstaniek replied:
This has been solved by doing all the calculation in UTC.

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+        if now < self.before:
+            return self.before
+
+        self._turn_to_next_day()
```
A property access shouldn't modify anything IMO

kstaniek replied:
Will find another way. It is confusing.

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,221 @@
+
+        async_track_point_in_time(
+            self.hass, self.point_in_time_listener,
+            self.next_change)
```
Also override `should_poll` and return False.

kstaniek replied:
ACK

frenck commented:
Could not find a matching documentation PR in our documentation repository, adding `docs-missing` label.

kstaniek committed:
Code refactored according to reviewer's suggestions

kstaniek commented:
Documentation added.

frenck commented:
Thanks for opening that PR @kstaniek 👍

kstaniek commented:
Guys, I've made all the changes according to the suggestions. Refactored the code. Changed the internal time to UTC to avoid DST problems. Added the test case for DST change. That's my first PR with new component. Really appreciate the reviewers' effort! @OttoWinter @dgomes @frenck

balloob commented:
I wonder if we're duplicating things that can be achieved with the calendar component.

kstaniek commented:
@balloob Paulus, with simple absolute time boundaries yes, but if you add sun events with offsets than every day the time ranges are different. I also solve the problem which I had with template based sensors, that after Hass restarts sometimes the sensors were transiting from unknown to on/off because the sun.sun component was not ready yet, I guess. As I wrote at one of my comment above I'm also planing to add 'analog' sensor with values calculated as normal distribution. So in the middle of the range the value will be 1 and on the edges will be 0 (bell curve). This could be interesting to use with ML and data analysing when you have to categorise time ranges.

kstaniek commented:
Guys, do you see any sense of adding:

```
self._unique_id = "{}-{}-{}-{}".format(
            after, after_offset.total_seconds(),
            before, before_offset.total_seconds())
```
and corresponding unique_id property?
The assumption is that if the time ranges are unique than the entity is unique.

balloob commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,250 @@
+        vol.Lower, cv.sun_event)),
+    vol.Optional(CONF_BEFORE_OFFSET, default=timedelta(0)): cv.time_period,
+    vol.Optional(ATTR_FRIENDLY_NAME): cv.string,
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
```
You need to use the `CONF_` constants, so CONF_NAME, CONF_ENTITY_ID

kstaniek replied:
Replaced ATTR_ with CONF_

balloob commented:
The problem with adding that unique ID is that changing the name will change the unique ID, and then you cannot rename the entity ID to the old one because it has been taken.

If you add config entry support, you can use config entry ID as a unique ID.

kstaniek commented:
Ok. I just realised I don't use friendly name attribute later one in the code. How should I implement it in a proper way?. The reason I'm asking for unique id is obviously that you can easily change the name later on in GUI. But in fact using this kind of sensor will rather very uncommon to present directly on the GUI. It's rather supportive sensor to simplify the automations.

kstaniek committed:
Fixed config template with friendly name support

kstaniek committed:
Finall pep8 fixes

OttoWinter commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,250 @@
+    def __init__(self, hass, name, friendly_name, after, after_offset,
+                 before, before_offset):
+        """Init the ToD Sensor..."""
+        self.hass = hass
```
Rather than passing in hass, please override `async_added_to_hass` and set up the initial values there.

kstaniek replied:
Moved the value initialisation to async_added_to_hass. Removed passing in hass.

OttoWinter reviewed (commented):
No, the HA 0.86.0 beta has already been cut off - plus this PR has unresolved comments.

For configuration format, please see https://github.com/home-assistant/home-assistant/pull/20068#issuecomment-453919773

kstaniek committed:
Removed async_generate_entity_id and moved initial calculation to async_added_to_hass

kstaniek committed:
Change the configuration schema to follow the current requirements

kstaniek commented:
It's ready to be merged now.

kstaniek commented:
Guys, any chance to merge eventually?

kstaniek commented:
Team, any update, please?

fabaff commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,234 @@
+ATTR_NEXT_UPDATE = 'next_update'
+
+
+SENSOR_SCHEMA = vol.Schema({
```
This is not used.

kstaniek replied:
Sure. Fixed

fabaff commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,234 @@
+    vol.Required(CONF_BEFORE): vol.Any(cv.time, vol.All(
+        vol.Lower, cv.sun_event)),
+    vol.Optional(CONF_BEFORE_OFFSET, default=timedelta(0)): cv.time_period,
+    vol.Required(CONF_NAME): cv.string
```
Please keep it ordered.

kstaniek replied:
@fabaff Ordered by what? Don't understand the reason. Should I put CONF_NAME first? What is the implication?

kstaniek replied:
I moved the CONF_NAME to the beginning. CONF_AFTER logically is first. Because the sensor is "ON" AFTER certain point in the time and BEFORE another point in the time. Offsets are logically  belonging to corresponding AFTER and BEFORE. Maybe there is another criteria I don't understand.

fabaff commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,234 @@
+
+    @property
+    def is_on(self):
+        """Return True is senso is on."""
```
Typo

kstaniek replied:
Fixed

fabaff commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,234 @@
+"""
+from datetime import datetime, timedelta
+import logging
+import pytz
```
Please keep the imports grouped. `isort` can help here if you don't want to do it manually.

kstaniek replied:
OK Done.

fabaff commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,234 @@
+        return self._next_update
+
+    @property
+    def state_attributes(self):
```
Should be `device_state_attributes()`.

kstaniek replied:
Fixed

fabaff commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,234 @@
+        return {
+            CONF_AFTER: self.after.astimezone(
+                self.hass.config.time_zone).isoformat(),
+            CONF_BEFORE: self.before.astimezone(
```
This should be `ATTR_BEFORE`.

kstaniek replied:
OK. Understand. Updated.

fabaff commented on homeassistant/components/binary_sensor/tod.py:
```diff
@@ -0,0 +1,234 @@
+    def state_attributes(self):
+        """Return the state attributes of the sun."""
+        return {
+            CONF_AFTER: self.after.astimezone(
```
This should be `ATTR_AFTER`.

kstaniek replied:
Fixed

kstaniek commented:
@fabaff Hi Fabian, thanks for your review. I have one comment above which I don't understand. The rest is clear. Waiting for your response to push the final commit.

kstaniek committed:
Update according latest suggestsion

fabaff committed:
Fix typos and minor changes

fabaff committed:
Fix lint issue

MartinHjelmare commented on tests/components/binary_sensor/test_tod.py:
```diff
@@ -0,0 +1,839 @@
+    get_astral_event_date, get_astral_event_next)
+
+
+class TestBinarySensorTod(unittest.TestCase):
```
We prefer standalone pytest test functions for new tests.

kstaniek replied:
@MartinHjelmare Is that a problem now?  I missed that requirement in developer's guide. I was following the Template Sensor implementation. How could I know it's wrong? Do I need to take any action?

MartinHjelmare replied:
You're welcome to rewrite the tests to pytest test functions if you want. It's not required at this point since this PR is merged.

I commented so that both contributors and reviewers can take note for the future.

Our developers docs are not complete.
